fedimint-core/src/encoding/mod.rs

```diff
@@ -611,6 +614,139 @@ impl Decodable for Cow<'static, str> {
+///
+/// So clients that don't have a corresponding module, can read
+/// the `len_u64` and skip the amount of data specified in it.
+///
+/// This type makes it more convenient. It's possible to attempt
+/// to retry decoding after more modules become available by using
+/// [`DynRawFallback::redecode_raw`].
+///
+/// Notably this struct does not ignore any errors. It only skips
+/// decoding when the module decoder is not available.
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+pub enum DynRawFallback<T> {
+    Raw {
+        module_instance_id: ModuleInstanceId,
+        raw: Vec<u8>,
+    },
+    Decoded(T),
+}
+
+impl<T> DynRawFallback<T>
+where
+    T: Decodable + 'static,
+{
+    pub fn expect_decoded(self) -> T {
+        match self {
+            DynRawFallback::Raw { .. } => panic!("Expected decoded value"),
+            DynRawFallback::Decoded(v) => v,
+        }
+    }
+
+    pub fn expect_decoded_ref(&self) -> &T {
+        match self {
+            DynRawFallback::Raw { .. } => panic!("Expected decoded value"),
+            DynRawFallback::Decoded(v) => v,
+        }
+    }
+
+    /// Attempt to re-decode raw values with new set of of `modules`
+    ///
+    /// In certain contexts it might be necessary to try again with
+    /// a new set of modules.
+    pub fn redecode_raw(
+        self,
+        decoders: &ModuleDecoderRegistry,
+    ) -> Result<Self, crate::encoding::DecodeError> {
+        Ok(match self {
+            DynRawFallback::Raw {
+                module_instance_id,
+                raw,
+            } => match decoders.get(module_instance_id) {
+                Some(decoder) => DynRawFallback::Decoded(decoder.decode(
+                    &mut &raw[..],
+                    module_instance_id,
+                    decoders,
+                )?),
+                None => DynRawFallback::Raw {
+                    module_instance_id,
+                    raw,
+                },
+            },
+            DynRawFallback::Decoded(v) => DynRawFallback::Decoded(v),
+        })
+    }
+}
+
+impl<T> From<T> for DynRawFallback<T> {
+    fn from(value: T) -> Self {
+        Self::Decoded(value)
+    }
+}
+
+impl<T> Decodable for DynRawFallback<T>
+where
+    T: Decodable + 'static,
+{
+    fn consensus_decode<R: std::io::Read>(
+        reader: &mut R,
+        decoders: &ModuleDecoderRegistry,
+    ) -> Result<Self, crate::encoding::DecodeError> {
+        let module_instance_id =
+            fedimint_core::core::ModuleInstanceId::consensus_decode(reader, decoders)?;
+        Ok(match decoders.get(module_instance_id) {
+            Some(decoder) => {
+                DynRawFallback::Decoded(decoder.decode(reader, module_instance_id, decoders)?)
+            }
+            None => {
+                // since the decoder is not available, just read the raw data
+                let total_len = u64::consensus_decode(reader, decoders)? as usize;
+                let mut buf = Vec::with_capacity(512);
```

Comment:
Shouldn't the capacity be `min(512, total_len)` in case `total_len < 512`?

Reply:
Thank for reminding me. After #2766 I should just use `Vec::consensus_decode` here.